remote/bitbucketserver/bitbucketserver.go

```diff
@@ -0,0 +1,358 @@
+}
+
+func (bs *BitbucketServer) Auth(token, secret string) (string, error) {
+	log.Info("Staring to auth for bitbucketServer. %s", token)
+	if len(token) == 0 {
+		return "", fmt.Errorf("Hasn't logged in yet")
+	}
+	return token, nil;
+}
+
+func (bs *BitbucketServer) Repo(u *model.User, owner, name string) (*model.Repo, error){
+	log.Info("Staring repo for bitbucketServer with user " + u.Login + " " + owner + " " + name )
+
+	client := NewClientWithToken(bs.ConsumerKey, bs.URL, u.Token)
+
+	url := bs.URL + "/rest/api/1.0/projects/" + owner + "/repos/" + name
+	log.Info("Trying to get " + url)
+	response, err := client.Get(url)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer response.Body.Close()
+	contents, err := ioutil.ReadAll(response.Body)
+	bsRepo := BSRepo{}
+	json.Unmarshal(contents, &bsRepo)
+
+	cloneLink := ""
+	repoLink := ""
+
+	for _, item := range bsRepo.Links.Clone {
+		if item.Name == "http" {
+			cloneLink = item.Href
+		}
+	}
+	for _, item := range bsRepo.Links.Self {
+		if item.Href != "" {
+			repoLink = item.Href
+		}
+	}
+	//TODO: get the real allow tag+ infomration
+	repo := &model.Repo{}
+	repo.Clone = cloneLink
+	repo.Link = repoLink
+	repo.Name=bsRepo.Slug
+	repo.Owner=bsRepo.Project.Key
+	repo.AllowTag=false
+	repo.AllowDeploy=false
+	repo.AllowPull=false
+	repo.AllowPush=true
+	repo.FullName = bsRepo.Project.Key +"/" +bsRepo.Slug
+	repo.Branch = "master"
+	repo.Kind = model.RepoGit
+
+
+	return repo, nil;
+}
+
+
+func (bs *BitbucketServer) Repos(u *model.User) ([]*model.RepoLite, error){
+	log.Info("Staring repos for bitbucketServer " + u.Login)
+	var repos = []*model.RepoLite{}
+
+	client := NewClientWithToken(bs.ConsumerKey, bs.URL, u.Token)
+
+	response, err := client.Get(bs.URL + "/rest/api/1.0/repos?limit=10000")
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer response.Body.Close()
+	contents, err := ioutil.ReadAll(response.Body)
+	var repoResponse Repos
+	json.Unmarshal(contents, &repoResponse)
+
+	for _, repo := range repoResponse.Values {
+		repos = append(repos, &model.RepoLite{
+			Name:     repo.Slug,
+			FullName: repo.Project.Key + "/" +  repo.Slug,
+			Owner: repo.Project.Key,
+		})
+	}
+
+
+	return repos, nil;
+}
+
+func (bs *BitbucketServer) Perm(u *model.User, owner, repo string) (*model.Perm, error){
+
+	//TODO: find the real permissions
+	log.Info("Staring perm for bitbucketServer")
+	perms := new(model.Perm)
+	perms.Pull = true
+	perms.Admin = true
+	perms.Push = true
+	return perms , nil
+}
+
+func (bs *BitbucketServer) File(u *model.User, r *model.Repo, b *model.Build, f string) ([]byte, error){
+	log.Info(fmt.Sprintf("Staring file for bitbucketServer login: %s repo: %s buildevent: %s string: %s",u.Login, r.Name, b.Event, f))
+
+	client := NewClientWithToken(bs.ConsumerKey, bs.URL, u.Token)
+	fileURL := fmt.Sprintf("%s/projects/%s/repos/%s/browse/%s?raw", bs.URL,r.Owner,r.Name,f)
+	log.Info(fileURL)
+	response, err := client.Get(fileURL)
+	if err != nil {
+		log.Fatal(err)
+	}
```

Comment:
`log.Fatal` will invoke `os.Exit(1)` and kill the server. Instead use `log.Error`